diff --git a/lib/CodeGen/AsmPrinter/WinException.cpp b/lib/CodeGen/AsmPrinter/WinException.cpp
index 9d7c96a1b8e..9633d4f382b 100644
--- a/lib/CodeGen/AsmPrinter/WinException.cpp
+++ b/lib/CodeGen/AsmPrinter/WinException.cpp
@@ -721,7 +721,7 @@ void WinException::emitCXXFrameHandler3Table(const MachineFunction *MF) {
   OS.EmitIntValue(0, 4);
 
   AddComment("EHFlags");
-  OS.EmitIntValue(1, 4);
+  OS.EmitIntValue(0, 4);
 
   // UnwindMapEntry {
   //   int32_t ToState;
diff --git a/lib/LTO/LTOCodeGenerator.cpp b/lib/LTO/LTOCodeGenerator.cpp
index 6af31e61f94..d8f8f703f82 100644
--- a/lib/LTO/LTOCodeGenerator.cpp
+++ b/lib/LTO/LTOCodeGenerator.cpp
@@ -423,6 +423,7 @@ void LTOCodeGenerator::applyScopeRestrictions() {
     MangledName.clear();
     MangledName.reserve(GV.getName().size() + 1);
     Mang.getNameWithPrefix(MangledName, &GV, /*CannotUsePrivateLabel=*/false);
+    if (MangledName == "_DllMain") return true;
     return MustPreserveSymbols.count(MangledName);
   };
 
